File: CovForDistributedSystems/CppAppWithUdpProbes/fractal_demo.cpp
#include <cmath>
#include <cstdint>
#include <cstdio>
#include <vector>
#include <string>
#include <fstream>
#include <chrono>
#include <thread>

#include "mprewriter.hpp"

static void write_ppm(const std::string& path, int w, int h, const std::vector<uint8_t>& rgb) {
    std::ofstream out(path, std::ios::binary);
    out << "P6\n" << w << " " << h << "\n255\n";
    out.write(reinterpret_cast<const char*>(rgb.data()), static_cast<std::streamsize>(rgb.size()));
}

int main() {
    mpr_start_sender();
    log_message("fractal_demo: starting Mandelbrot render");

    const int W = 800;
    const int H = 600;
    const int MAX_IT = 256;
    std::vector<uint8_t> rgb(W * H * 3);

    // Outer scope probe for the whole render
    mprewriter_scope_START(2000);

    for (int y = 0; y < H; ++y) {
        // Per-row probe
        mprewriter_scope_START(2001);
        double cy = (y - H / 2.0) * 4.0 / H; // map to [-2,2]
        for (int x = 0; x < W; ++x) {
            // Per-pixel probe
            mprewriter_scope_START(2002);
            double cx = (x - W / 2.0) * 4.0 / W;
            double zx = 0.0, zy = 0.0;
            int it = 0;
            // Iteration probe
            {
                mprewriter_scope_START(2003);
                while (zx*zx + zy*zy < 4.0 && it < MAX_IT) {
                    double xt = zx*zx - zy*zy + cx;
                    zy = 2.0*zx*zy + cy;
                    zx = xt;
                    ++it;
                }
            }
            // Simple palette
            int idx = (y * W + x) * 3;
            uint8_t c = static_cast<uint8_t>(it % 256);
            rgb[idx + 0] = c;
            rgb[idx + 1] = (c * 5) % 256;
            rgb[idx + 2] = (c * 13) % 256;
        }
        if (y % 50 == 0) log_message("fractal_demo: row " + std::to_string(y));
    }

    write_ppm("fractal.ppm", W, H, rgb);
    log_message("fractal_demo: wrote fractal.ppm");

    // Flush and shutdown
    std::this_thread::sleep_for(std::chrono::milliseconds(50));
    close_probe();
    mpr_join_sender();
    return 0;
}

File: CovForDistributedSystems/RustAppWithUdpProbes/udp_probe_demo/app/src/fractal_demo.rs
use crate::mprewriter;
use mprewriter_macros::mprewriter_scope_START;

pub fn run_demo() {
    // Context label for grouping
    mprewriter::set_context("fractal_demo::run_demo");
    mprewriter_scope_START!(2000); // outer render scope
    mprewriter::log_message("Fractal demo: starting Mandelbrot render");

    const W: usize = 800;
    const H: usize = 600;
    const MAX_IT: u32 = 256;

    let mut rgb = vec![0u8; W * H * 3];

    for y in 0..H {
        mprewriter_scope_START!(2001); // per-row scope
        let cy = (y as f64 - (H as f64) / 2.0) * 4.0 / (H as f64);
        for x in 0..W {
            mprewriter_scope_START!(2002); // per-pixel scope
            let cx = (x as f64 - (W as f64) / 2.0) * 4.0 / (W as f64);
            let mut zx = 0.0f64;
            let mut zy = 0.0f64;
            let mut it: u32 = 0;
            {
                mprewriter_scope_START!(2003); // iteration scope
                while zx * zx + zy * zy < 4.0 && it < MAX_IT {
                    let xt = zx * zx - zy * zy + cx;
                    zy = 2.0 * zx * zy + cy;
                    zx = xt;
                    it += 1;
                }
            }
            let idx = (y * W + x) * 3;
            let c = (it % 256) as u8;
            rgb[idx] = c;
            rgb[idx + 1] = c.wrapping_mul(5);
            rgb[idx + 2] = c.wrapping_mul(13);
        }
        if y % 50 == 0 {
            mprewriter::log_message(&format!("fractal_demo: row {}", y));
        }
    }

    // Write PPM
    if let Err(e) = write_ppm("fractal_rust.ppm", W as u32, H as u32, &rgb) {
        mprewriter::log_message(&format!("fractal_demo: failed to write PPM: {}", e));
    } else {
        mprewriter::log_message("fractal_demo: wrote fractal_rust.ppm");
    }
}

fn write_ppm(path: &str, w: u32, h: u32, rgb: &[u8]) -> std::io::Result<()> {
    use std::io::Write;
    let mut f = std::fs::File::create(path)?;
    write!(f, "P6\n{} {}\n255\n", w, h)?;
    f.write_all(rgb)?;
    Ok(())
}
